### PR TITLE
FetchBaseSVID/FetchSVID refactor

### LINKED ISSUE #464: SPIRE server should inform agents of _all_ registration entries
Currently, when the agent calls the SPIRE server, the server uses the SPIFFE ID in the mTLS client certificate in order to authorize CSR requests. The authorization logic looks one level deep, asserting that the requestor is a parent of the entry containing the SPIFFE ID for which it is trying to get a CSR signed.

Unfortunately, this leads to a lot of complication in the agent. The agent must first make a request with its own certificate and get all the direct descendants signed. If there are any chained entries, it must then make subsequent requests using the newly-received credentials in order to obtain further IDs.

Update the SPIRE server to perform recursion over the registration entries, such that it can determine every ID that a requestor is _eventually_ authorized to issue. Return to the agent all of these IDs, allowing it to create CSRs for everything at once. Also update the authorization in the CSR signing logic to authorize operations for the full set of IDs, as opposed to the immediate descendants. Update the agent's cache manager so that it no longer has to manage multiple connections with multiple SVIDs.

### PR DESCRIPTION
**Pull request by azdagron**
This PR changes the registration entries returned by FetchBaseSVID and FetchSVID. Instead of returning just the entries directly authorized for the node, it now does a recursive descent through all descendent registry entries as well as those reachable via node resolver mappings.

The agent code is drastically reduced since it can make a single call to retrieve all registration entries it is authorized to issue, and then turn around and request all CSRs in a single call. The client pool is no longer required, nor special case handling for vanity entries.

fixes #464

### CODE REVIEW

File: pkg/common/util/sort.go [hunk 1]
```diff
@@ -1,14 +1,112 @@
...
+func cloneRegistrationEntries(entries []*common.RegistrationEntry) []*common.RegistrationEntry {
+	cloned := make([]*common.RegistrationEntry, 0, len(entries))
+	for _, entry := range entries {
+		cloned = append(cloned, cloneRegistrationEntry(entry))
+	}
+	return cloned
+}
+
+func cloneRegistrationEntry(entry *common.RegistrationEntry) *common.RegistrationEntry {
+	return proto.Clone(entry).(*common.RegistrationEntry)
```
Comment (evan2645) [inline pkg/common/util/sort.go]: nice

File: pkg/server/endpoints/node/handler.go [hunk 2]
```diff
@@ -624,3 +570,114 @@ func getSpiffeIDFromCSR(csr []byte) (spiffeID string, err error) {
 
 	return spiffeID, nil
 }
+
+func FetchRegistrationEntries(ctx context.Context,
```
Comment (evan2645) [inline pkg/server/endpoints/node/handler.go]: Maybe this code would be better placed in a dedicated file? `entries.go`, `fetch_entries.go`, or something similar?

File: pkg/server/endpoints/node/handler_test.go [hunk 3]
```diff
@@ -158,60 +164,90 @@ func TestFetchSVIDWithRotation(t *testing.T) {
...
-		{
-			Type:  "a",
-			Value: "1",
-		},
-		{
-			Type:  "b",
-			Value: "2",
-		},
+	assert := assert.New(t)
+	dataStore := fakedatastore.New()
```
Comment (evan2645) [inline pkg/server/endpoints/node/handler_test.go]: Thank you for doing this ❤️

File: pkg/server/endpoints/node/handler.go (hunk 2 above)
Comment (azdagron) [reply pkg/server/endpoints/node/handler.go]: good idea. i intended to move it into its own package (could be useful to cmdline utilities or other packages outside of the node handler). I'll move it.

**Pull request merged**